### PR TITLE
Add ability to read messages with pagination

### PR DESCRIPTION
**Pull request by neekolas**
## Summary
Adds ability to read paginated messages, resolving https://github.com/xmtp/xmtp-js/issues/168.

These are the options I have made available to control the query:
```ts
export type ListMessagesPaginatedOptions = {
  startTime?: Date
  endTime?: Date
  pageSize?: number
  direction?: messageApi.SortDirection
}
```

## Notes
- This returns an `AsyncGenerator` yielding one page of results at a time. While I could simply return an iterator of all messages instead of an iterator of pages and still use pagination under the hood, this is a little more performant (saves waiting until `process.nextTick()` for each `yield`/message). It also lets the caller process the results one page at a time, which may be advantageous for React apps where they would want to batch updates to the DOM. I could be convinced to go the other way.
- I have not implemented listing conversations in a paginated manner. Mostly because no one has ever asked for that. Maybe we want to offer both just for the sake of consistency (it's trivially easy), but for the moment I'm inclined to punt for the sake of simplicity.

### CODE REVIEW

File: src/Client.ts [hunk 1]
```diff
@@ -50,6 +51,13 @@ export type ListMessagesOptions = {
   limit?: number
 }
 
+export type ListMessagesPaginatedOptions = {
+  startTime?: Date
+  endTime?: Date
+  pageSize?: number
+  direction?: messageApi.SortDirection
```
Comment (mkobetic) [inline src/Client.ts:58]: I guess we don't want limit here, because you can just quit the pagination loop whenever you want, correct?

**Comment by mkobetic** [comment]
This looks good, but one thing I wonder is that we're using Streams for new messages and async generator for pages of historic messages. The two constructs seem kinda equivalent to me, are they? If so using both may make the API look a bit schizophrenic. Is that a valid concern?

File: src/Client.ts (hunk 1 above)
Comment (neekolas) [reply src/Client.ts:58]: That's right

**Comment by neekolas** [comment]
> The two constructs seem kinda equivalent to me, are they? If so using both may make the API look a bit schizophrenic. Is that a valid concern?

From the user's perspective, the Streams are really just async iterators. The whole Stream concept is mostly an internal class for us to manage filling the async iterator from callbacks. Async Iterators and Async Generators work the same from the caller's perspective, so I'm not sure most users will notice.

Having worked with Async Generators a bit on the `ApiClient`, I'm starting to think it would be a nice refactor to rewrite the Stream class as an Async Generator as well. The code is much [cleaner](https://github.com/xmtp/xmtp-js/blob/main/src/ApiClient.ts#L226) and it handles a lot of the semantics around returning from the stream for you natively.

**Comment by mkobetic** [comment]
> Having worked with Async Generators a bit on the `ApiClient`, I'm starting to think it would be a nice refactor to rewrite the Stream class as an Async Generator as well. The code is much [cleaner](https://github.com/xmtp/xmtp-js/blob/main/src/ApiClient.ts#L226) and it handles a lot of the semantics around returning from the stream for you natively.

Yeah, maybe the right thing to do is to get rid of the Stream in favour of the generator. Can we do that with reasonably backward compatible manner?

**Comment by neekolas** [comment]
> Can we do that with reasonably backward compatible manner?

Ya, that's the nice thing. Callers wouldn't even notice the change. Only the types would be different, but you can use them the same way.

**Comment by neekolas** [comment]
@mkobetic what do you think about merging this as-is and then I will make a backlog ticket for converting our Stream implementation to an Async Generator

**Review by mkobetic — APPROVED** [review]
Yes, :shipit:. Sorry I thought I already 👍 this.

**Pull request merged**